packages/babel-register/src/loader.mjs

```diff
@@ -0,0 +1,287 @@
+
+    if (result !== null) {
+      return result;
+    }
+  }
+
+  // Let Node.js handle all other URLs.
+  return nextLoad(url);
+}
+
+export function globalPreload({ port }) {
+  port.onmessage = evt => {
+    if (evt.data.opts) {
+      global._BABEL_ESM_REGISTER.register(evt.data.client, evt.data.opts);
+    } else if (evt.data.revert) {
+      global._BABEL_ESM_REGISTER.revert();
+    }
+  };
+  port.unref();
+
+  const fnc = {
+    register(client, opts) {
+      port.postMessage({ client, opts });
+    },
+    revert() {
+      port.postMessage({ revert: true });
+    },
+  };
+  return `\
+    globalThis._BABEL_ESM_REGISTER = { ${fnc.register.toString()}, ${fnc.revert.toString()} };
+    port.unref();
+  `;
+}
+
+export function getGlobalPreloadCode() {
+  const fnc = {
+    register() {
+      throw new Error(
+        "Modyfing @babel/register on-demand is not supported in this version of node"
+      );
+    },
+    revert() {
+      throw new Error(
+        "Modyfing @babel/register on-demand is not supported in this version of node"
+      );
+    },
+  };
+  return `\
+    globalThis._BABEL_ESM_REGISTER = { ${fnc.register.toString()}, ${fnc.revert.toString()} };
+  `;
+}
+
+// backwards compat
+export async function getSource(url, context, next) {
+  return load(url, context, next);
+}
+
+export async function getFormat(url, context, next) {
+  if (flag[url]) {
+    const format = await getPackageType(url);
+    return {
+      format: format || "commonjs",
+    };
+  }
+  return next(url, context, next);
+}
+
+async function getPackageType(url) {
+  // `url` is only a file path during the first iteration when passed the
+  // resolved url from the load() hook
+  // an actual file path from load() will contain a file extension as it's
+  // required by the spec
+  // this simple truthy check for whether `url` contains a file extension will
+  // work for most projects but does not cover some edge-cases (such as
+  // extensionless files or a url ending in a trailing space)
+  const ext = extname(url);
+  if (ext === ".cjs") return "commonjs";
+  if (ext === ".mjs") return "module";
+  const isFilePath = !!ext;
+  // If it is a file path, get the directory it's in
+  const dir = isFilePath ? dirname(fileURLToPath(url)) : url;
+  // Compose a file path to a package.json in the same directory,
+  // which may or may not exist
+  const packagePath = resolvePath(dir, "package.json");
+  // Try to read the possibly nonexistent package.json
+  const type = await readFile(packagePath, { encoding: "utf8" })
+    .then(filestring => JSON.parse(filestring).type)
+    .catch(err => {
+      if (err?.code !== "ENOENT") console.error(err);
+    });
+  // Ff package.json existed and contained a `type` field with a value, voila
+  if (type) return type;
+  // Otherwise, (if not at the root) continue checking the next directory up
+  // If at the root, stop and return false
+  return dir.length > 1 && getPackageType(resolvePath(dir, ".."));
```

Comment:
this is likely going to break on windows, since dir will end up being d: or something. is there a function to detect if we have a root folder for a certain drive letter? or just dir.split(path.sep).length === 1?